### PR TITLE
Rework serialization of XMPPAccount and ColorIDSet

### PR DESCRIPTION
**Pull request by m273d15**
* Cleaned up the tests of XMPPAccountStore
* Used XStream in order to (de)serialize XMPPAccount and ColorIDSet
#491

### CODE REVIEW

File: core/src/saros/account/XMPPAccountStore.java [hunk 1]
```diff
@@ -224,31 +200,18 @@ private synchronized void saveAccounts() {
...
-
-      byte[] activeAccount = Crypto.encrypt(out.toByteArray(), secretKey);
-      out.reset();
-
-      oos = new ObjectOutputStream(out);
-      oos.writeObject(accounts);
-      oos.flush();
+      dataOut = new FileOutputStream(accountFile);
+      // use a pair in order to create a artificial xml root node
+      xStream.toXML(new ImmutablePair(activeAccount, accounts), out);
```
Comment (srossbach) [inline core/src/saros/account/XMPPAccountStore.java]: List<XMPPAcounts> accountsToSave = new ArrayList<>(accounts)
int activeAccountIdx = accountsToSave.indexOf(acivteAccount)

new ImmutablePair(activeAccountIdx, accountsToSave)

File: core/src/saros/account/XMPPAccountStore.java [hunk 2]
```diff
@@ -594,7 +566,7 @@ public XMPPAccount findAccount(String jidString) {
    * save the result in the user home directory which should be default only be accessible by the
    * user itself or administrators.
    */
-  private static class Crypto {
+  static class Crypto {
```
Comment (srossbach) [inline core/src/saros/account/XMPPAccountStore.java]: Why did you remove private ?

File: core/src/saros/account/XMPPAccountStore.java [hunk 3]
```diff
@@ -150,44 +149,21 @@ private synchronized void loadAccounts() {
...
-              new ObjectInputStream(new ByteArrayInputStream(Crypto.decrypt(buffer, secretKey)))
-                  .readObject();
-
-      size = dataIn.readInt();
+      dataIn = new FileInputStream(accountFile);
 
-      if (size <= 0 || size > MAX_ACCOUNT_DATA_SIZE)
-        throw new IOException(
-            "account data seems malformed, refused to load " + size + " bytes into memory");
+      byte[] encryptedAccountData = IOUtils.toByteArray(dataIn);
```
Comment (srossbach) [inline core/src/saros/account/XMPPAccountStore.java]: So what happens if you point to a garbage file with about several 100 MBytes in size ?

File: core/src/saros/account/XMPPAccountStore.java [hunk 4]
```diff
@@ -264,6 +227,15 @@ private synchronized void saveAccounts() {
     LOG.debug("saved " + accounts.size() + " account(s)");
   }
 
+  private XStream createXStream() {
+    XStream xStream = new XStream();
+    xStream.alias("accounts", ImmutablePair.class);
+    xStream.aliasField("activeAccount", ImmutablePair.class, "left");
```
Comment (srossbach) [inline core/src/saros/account/XMPPAccountStore.java]: Isn't there another way to retrieve the field name left & right in a compile safe fashion ?

File: core/src/saros/editor/colorstorage/ColorIDSetStorage.java [hunk 5]
```diff
@@ -182,12 +177,15 @@ private void load() {
...
-              new ObjectInputStream(new ByteArrayInputStream(serializedData.getBytes(CHARSET)))
-                  .readObject();
+      currentAvailableSets = (List<ColorIDSet>) createXStream().fromXML(serializedData);
     } catch (Exception e) {
       LOG.error("error while loading color sets", e);
     }
   }
+
+  private XStream createXStream() {
+    XStream xStream = new XStream(new DomDriver(CHARSET));
```
Comment (srossbach) [inline core/src/saros/editor/colorstorage/ColorIDSetStorage.java]: Is this the same behavior as before ? To be more specific, is there any need for a charset at all ? Because you are producing already a string.

File: core/src/saros/account/XMPPAccountStore.java (hunk 1 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: I like the idea. This will simplify the XML structure (without a reference)

File: core/src/saros/account/XMPPAccountStore.java (hunk 2 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: For testing purposes. Otherwise it is not possible to test the accountFile format.

File: core/src/saros/account/XMPPAccountStore.java (hunk 3 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: I will add a check whether the file is smaller than MAX_ACCOUNT_DATA_SIZE

File: core/src/saros/account/XMPPAccountStore.java (hunk 4 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: It seems as if xstream does not support a compile safe way.

File: core/src/saros/editor/colorstorage/ColorIDSetStorage.java (hunk 5 above)
Comment (m273d15) [reply core/src/saros/editor/colorstorage/ColorIDSetStorage.java]: I will check this. Thank you, for the review

Comment (m273d15) [reply core/src/saros/editor/colorstorage/ColorIDSetStorage.java]: Done

File: core/src/saros/account/XMPPAccountStore.java (hunk 3 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: Done

File: core/src/saros/account/XMPPAccountStore.java (hunk 1 above)
Comment (m273d15) [reply core/src/saros/account/XMPPAccountStore.java]: Done

File: core/test/junit/saros/account/XMPPAccountStoreTest.java [hunk 6]
```diff
@@ -89,26 +89,16 @@ private void writeAccountFile(File accountFile, String key, String content) thro
   }
 
   private String createAccountFileContent(
-      XMPPAccount activeAccount, List<XMPPAccount> otherAccounts) {
+      XMPPAccount activeAccount, ArrayList<XMPPAccount> configuredAccounts) {
```
Comment (srossbach) [inline core/test/junit/saros/account/XMPPAccountStoreTest.java]: I do not get idea behind this test. Do you not trust XStream ? What is wrong with creating an XMPPAccountStore, add some accounts and save it. Afterwards you create another XMPPAccountStore and use the existing file. I guess such a test already exists.

File: core/test/junit/saros/account/XMPPAccountStoreTest.java [hunk 7]
```diff
@@ -89,26 +89,16 @@ private void writeAccountFile(File accountFile, String key, String content) thro
...
 
+    Integer index = configuredAccounts.indexOf(activeAccount);
```
Comment (srossbach) [inline core/test/junit/saros/account/XMPPAccountStoreTest.java]: It is "int" and not "integer" unless you want to return null. I have seen this multiple times in this patch set.

File: core/test/junit/saros/account/XMPPAccountStoreTest.java (hunk 6 above)
Comment (m273d15) [reply core/test/junit/saros/account/XMPPAccountStoreTest.java]: The idea was to inform a developer if he/she changes the XML format/content of the file and therefore invalidates our old files.

File: core/test/junit/saros/account/XMPPAccountStoreTest.java (hunk 7 above)
Comment (m273d15) [reply core/test/junit/saros/account/XMPPAccountStoreTest.java]: Done

File: core/test/junit/saros/account/XMPPAccountStoreTest.java (hunk 6 above)
Comment (srossbach) [reply core/test/junit/saros/account/XMPPAccountStoreTest.java]: You can also use Aliases to allow variable renaming in the future. And I also think adding a comment to the affected classes (e.g if you change XYZ you will broke ABC) is much more helpful than only have one test case that is failing.

Comment (m273d15) [reply core/test/junit/saros/account/XMPPAccountStoreTest.java]: I will add these additional information to the corresponding lines.

Comment (m273d15) [reply core/test/junit/saros/account/XMPPAccountStoreTest.java]: Done
